### PR TITLE
feat: add custompick component (#2855)

### LINKED ISSUE #2855: extension-api: Add a choice menu / generic preset
### Is your enhancement related to a problem? Please describe

OpenShift extension need to display preset with nice UI

### Describe the solution you'd like

Implements an API where we can list items to make a choice more graphical

### Describe alternatives you've considered

_No response_

### Additional context

image

### PR DESCRIPTION
**Pull request by lstocchi**
### What does this PR do?

This PR adds a custompick component so extensions can show a options in a custom dialog.

### Screenshot/screencast of this PR

when having 2 items
two_items_full_length

with >3
custompick_draggable

### What issues does this PR fix or reference?

it resolves #2855 

### How to test this PR?

1. install quay.io/lstocchi/sample-podman-extension:custompick_v3
2. click on the custompick item in the statusbar
3. select an option and check that it's index is printed in the console

P.S: Adding some tests but it's ready if you want to give a look

### CODE REVIEW

**Comment by deboer-tim** [comment]
Going back to the original issue, it's not clear to me how the OpenShift Local extension would use this long term. Today it is a simple popup, but we have talked about this choice being embedded in an onboarding flow, or in Settings > Resources if you do it later. If you want to change it later it's a preference today, and I don't know how this would fit in to changing it there.

In terms of the UI, the look is good but this is falling into an odd place between a quickpick (a quick selection at top of screen) and message box (normal 'dialog' with close button). I think we need to push it to look more like one or the other, e.g. if we did the latter, make the title bar match message box and give it a close button.

The Show Details resizing and moving the dialog is also a bit distracting to me. I don't know if @mairin has any ideas, but I'm wondering if we should use scrollbars or keep the top fixed or something.

**Comment by mairin** [comment]
@deboer-tim my thought was that it'd be embedded in the settings > resources area as another page instaed of a dialog, eg 
Screenshot from 2023-06-26 11-50-49

If it was a page it wouldn't have the same size changing issues I think?

But yeh it's not something that can be changed once the openshift local is setup/installed, so it's really designed as an onboarding step page. but i suspect this PR is more about making the framework to have a selection component like this and less about the specific openshift  local workflow?

@lstocchi i am amazed how accurately you were able to reproduce the design in code!!! it looks really nice :)

**Comment by lstocchi** [comment]
> Going back to the original issue, it's not clear to me how the OpenShift Local extension would use this long term. Today it is a simple popup, but we have talked about this choice being embedded in an onboarding flow, or in Settings > Resources if you do it later. If you want to change it later it's a preference today, and I don't know how this would fit in to changing it there.

I worked on a dialog-style solution after talking with Florent so i think he can better explain his thought on the usage. Maybe it's temporary replace for the openshift local and it will be used in other use cases?? @benoitf 

 > i am amazed how accurately you were able to reproduce the design in code!!! it looks really nice :)

Thanks!! it's easy when you have a mockup to copy 😆

**Comment by mairin** [comment]
Maybe to better support it in a dialog format... can we make the window height by default the height of the tallest one expanded? :-/ ?

File: packages/renderer/src/lib/dialogs/CustomPick.svelte [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+        {#if title}
+          <h1 class="grow text-lg font-bold capitalize" aria-label="{title}">{title}</h1>
+        {/if}
+        {#if description}
+          <h3 class="grow text-xs">{description}</h3>
+        {/if}
+      </div>
+
+      <div class="flex flex-row flex-wrap mx-auto">
+        {#each items as item, i}
```
Comment (benoitf) [inline packages/renderer/src/lib/dialogs/CustomPick.svelte]: FYI we're not doing it yet in our components, so mainly a FYI but we should add keys
https://svelte.dev/tutorial/keyed-each-blocks

**Comment by benoitf** [comment]
how does it work when we have more than 2 columns or only one column or no column ?

**Review by benoitf — APPROVED** [review]
works for me (you can use @mairin feedback)

only think wired is the 'next' button if we're just asking a user for a single thing. I'm not sure I would expect 'next' there (more `OK` or `Done` or something else)

**Comment by lstocchi** [comment]
> how does it work when we have more than 2 columns or only one column or no column ?

More than 2 cols, it puts 2 elements per row and adds a scrollbar. Idk if this has to be changed somehow (e.g with 3 options, put all three in one row and reduce their width? @mairin ). I'll add a gif shortly

multiple-items
P.S: just noticed the margin on the right is larger then the one on the left with multiple items. Fixing it

Maybe it's clearer this way if there are odds items?
three-items


No column -> it only shows the title, description. I think it's up to the creator to add atleast an item, no? Do i have to show a message when this happen or prevent from displaying the dialog?

File: packages/renderer/src/lib/dialogs/CustomPick.svelte (hunk 1 above)
Comment (lstocchi) [reply packages/renderer/src/lib/dialogs/CustomPick.svelte]: by looking at the example it uses the property id of the item as a key. In my case i don't have anything like that. I suppose i should use the array index or the entire item object? `{#each items as item, i (i)}`

**Comment by lstocchi** [comment]
> Maybe to better support it in a dialog format... can we make the window height by default the height of the tallest one expanded? :-/ ?

@mairin The main problem is how it should behave with multiple items. 
My idea was to put the body/footer invisible instead of hidden so to maintain the space. And it works fine with few items (just don't look at the hover background color in the gif below, i have to fix it) 
two-items_fixed_height



.... but it's ugly when you have multiple rows. 
What should i do? Or any other way to do it instead of using `invisible`? How do you caluclate the expanded height without it being rendered? The item can be created collapsed at first (there is a flag the user can use)

three-items_fixed_height

**Comment by mairin** [comment]
@lstocchi hmmmmm 🤔 How much do we want this component to scale before we use another one? Up to 3, 4, 5?

I think if you can't get 3 columns.... revert from expand in place and have the additional info appear in some kind of layer that pops up on top? I might be explaining badly, let me include a bad ugly sketch 😅

Screenshot_20230627_120822_Samsung Notes

Hopefully this makes sense! Just layer on top basically?

**Comment by lstocchi** [comment]
> @lstocchi hmmmmm 🤔 How much do we want this component to scale before we use another one? Up to 3, 4, 5?

good question. @benoitf should we have any limit?

**Comment by lstocchi** [comment]
> if you can't get 3 columns

I can also put 3 cols in a row but the problem remains when having more than 3 columns

> Hopefully this makes sense! Just layer on top basically?

👍 I'll try creating something

**Comment by mairin** [comment]
@lstocchi sweet! Just to be clear - I would say keep expander if you got room for it, if thats 2 or 3 fine. Beyond that, make it pop on top instead of expand. (if possible.)

**Review by cdrage — APPROVED** [review]
Tested, works for me. Merge whenever you have confirmed UI changes with @mairin

**Comment by lstocchi** [comment]
@mairin here the new design. It took a while, sorry. WDYT? 
When the details use a popup, there is no limit of how many you can open. I thought that a user could want to compare the details of different items before to select one. So she has to close the details dialog with the `X`. 

when having 2 items
two_items_full_length

3 items
three_items_full_length


\>3 items
five_items_cards

**Comment by mairin** [comment]
@lstocchi this is awesome!! I just have two nits! :)

- With the 3 column one... is there any way to force the third one to be the same height as the first and second?
- With last one, where you have 5 and you have the floating pop ups? Is there any way to maybe have a little triangle like /\ at the top center so it's clear which card the popup belongs to?

Also a question... what happens if I want to compare #1 with #4? does the popup for #1 block the one for #4 or vice-versa?

**Comment by lstocchi** [comment]
> * With the 3 column one... is there any way to force the third one to be the same height as the first and second?

The problem is that we don't know the content length in advance. In this specific case i could add a `min-height` so to have everything with the same height but if the user adds a lot of content in one of the option we ends up having the same behavior.
Is it ok for you? Were you thinking at something else? Because if we use the height of the tallest element for all items than there could be cards with few content and a lot of empty space 

> * With last one, where you have 5 and you have the floating pop ups? Is there any way to maybe have a little triangle like /\ at the top center so it's clear which card the popup belongs to?

👍 

> Also a question... what happens if I want to compare #1 with #4? does the popup for #1 block the one for #4 or vice-versa?

Good point. Should i make the popups draggable so they can be moved if necessary?

**Comment by mairin** [comment]
On the height: would it be possible to give the person using these
components the ability to specify a min-height of their own choosing?

On the 2nd - draggable sounds like a good solution!

**Comment by lstocchi** [comment]
This is the final outcome.
Thanks @mairin for the help

custompick_draggable

**Comment by lstocchi** [comment]
> On the height: would it be possible to give the person using these components the ability to specify a min-height of their own choosing? 

Added it. Now users can set a min-height to their components

**Pull request merged**